### PR TITLE
Budget-App

### PR DESCRIPTION
**Pull request by MohammadYaser**
Added the Migration.
Added the Model.
Added the devise.
Added model test.
Fix N+1 problem

Sign up and log in pages

The user should be able to register in the app with full name, email and password (all mandatory).
The user can log into the app using email and password.
If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).
Home page (categories page)

When the user logs in, they are presented with the categories page.
For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
There is a button "add a new category" at the bottom that brings the user to the page to create a new category.
Transactions page

For a given category, the list of transactions is presented, ordered by the most recent.
At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
When the user clicks on the "Back" button (<), the user navigates to the home page.
"Add a new category" page

The user fills out a form to create a new category, indicating their name and icon (both mandatory).
The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
When the user clicks on the "Back" button (<), the user navigates to the home page.
"Add a new transaction" page

The user fills out a form to create a new transaction with:
name (mandatory)
amount (mandatory)
categories (mandatory at least one)
The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.
Testing requirements
Create unit and integration tests for all the most important components of your RoR application.
Technical requirements
You should use Postgres as your database.
You should use devise for authentication.
You should validate all user input to make sure that anyone with bad intentions cannot compromise your app.
You can use a view template engine of your choice (.erb, .slim, .haml).
The project should be deployed and accessible online.

### CODE REVIEW

File: app/views/home/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    
+    <!-- Footer section -->
+    <footer>
+        
+        <!-- Link to navigate to the login page -->
+        <%= link_to "LOGIN", new_user_session_path, id: "login" %>
+        
+        <!-- Link to navigate to the sign-up page -->
+        <%= link_to "SIGN UP", new_user_registration_path, id: "signup" %>
+    </footer>
```
Comment (tugsem) [inline app/views/home/index.html.erb:12]: - [ ] Please center the buttons and add your app logo on the top to match the given design: 

yours|given
---|---
image|image

File: app/views/categories/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        
+        <!-- Category name and creation date -->
+        <div>
+          <p class="name"><%= category.name %></p>
+          <p class="expense-date"><%= category.created_at.strftime("%m/%d/%Y") %></p>
+        </div>
+        
+        <!-- Total expense amount for the category -->
+        <div class="amount">
+          <p><%= number_to_currency(category.expenses.sum(:amount)) %></p>
```
Comment (tugsem) [inline app/views/categories/index.html.erb]: - [ ] Right now when I add new transactions the categories index page doesn't display the total number of transactions. It shows 0.00$ all the time. Please fix it.
image

File: README.md [hunk 3]
```diff
@@ -1 +1,135 @@
...
+**I would like to thank the following individuals and organizations for their contribution to this project.**
+
+- I would like to express my heartfelt gratitude to [**Microverse**](https://www.microverse.org/?grsf=mohammad-a-nbtazu) for the invaluable learning experience they have provided.
+
+Original design idea by [Gregoire Vella in Behance](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=).
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+## 📜 License <a name="license"></a>
+This project is [MIT](./LICENSE) licensed.
```
Comment (tugsem) [inline README.md:135]: - [ ] Please fix the license link as it is broken.
image

**Review by tugsem — COMMENTED** [review]
##  STATUS: REQUIRED CHANGES ♻️
Hello @MohammadYaser  👋 , 

Good job so far! 👏 
There are some issues that you still need to work on to go to the next project but you are almost there!

Alt Text
## Highlights 🚀 
✔️linters are passing
✔️correct use of gitflow
✔️live link and video presentation added successfully
✔️ app tested thoroughly

## Video Presentation
- The student recorded a video presenting the features of the project.
- The maximum length of the video is 5 minutes.
- The video contains a clear and objective description of the project in professional English. The topic and purpose of the project are made clear.
- The video contains a demo of the project, highlighting all the project features.
- The video contains a highlight about a piece of the code the student finds interesting or is proud of making.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/categories/index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    # Looping through each category and making assertions
+    @categories.each do |category|
+      expect(page).to have_selector('.category-link', count: @categories.length)
+
+      # Using within to scope the assertions to the specific category link
+      within(".category-link[href='#{user_category_expenses_path(@user, category)}']") do
+        expect(page).to have_selector(".category-icon img[src=\"#{category.icon}\"]")
+        expect(page).to have_selector('.category p', text: category.name)
+        expect(page).to have_selector('.category p.expense-date', text: category.created_at.strftime('%m/%d/%Y'))
+        expect(page).to have_selector('.amount p', text: number_to_currency(category.expenses.sum(:amount)))
```
Comment (OyePriscilla) [inline spec/features/categories/index_spec.rb]: - [ ] Please kindly take note that this test is failing, though I observed that it passed in the previous commit before the review. Kindly re-adjust the implementation; I think the reason for this is the manipulation you did to display the total amount for individual category. 🙏

**Review by OyePriscilla — CHANGES_REQUESTED** [review]
Hi @MohammadYaser 👍 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

so-close-so-close-seth-meyers

 ## Highlights
License link redirected appropriately ✔️ 
Designed sign-up page as suggested ✔️ 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by VICTOR-CHUKWU — APPROVED** [review]
Hi @MohammadYaser , 
# STATUS: APPROVED! 🥇 💯 
done

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- [ ] All linters passed
- [ ] All tests passed

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**